User context

darthjee opened the pull request:
In order to allow better extendability within the user cotext, this PR wraps it in a class with no methods or attributes that will behave just like ```Object``` but still allow for user to write helpers that extend it without affecting the behavior of ```Object``` objects

## Description
- [x] New UserContext class
- [x] Split extract extend method from within clone
- [x] Use clonedSharedUserContext when a clone of user context is needed

## Motivation and Context
While I was trying to write a method that acts just like let for ruby, I discovered that I couldn't easily extend the user context without ruining the behavior of Object class

## How Has This Been Tested?
It has mainly been tested through specs and by extending it in a spec on a separated project

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Tested on linux with Grunt
- [x] Tested on Linux Firefox (rake jasmine)
- [x] Tested on Linux Chrome (rake jasmine)
- [ ] Tested on windows internet explorer (rake jasmine)

slackersoft commented on src/core/Suite.js:
```diff
@@ -147,15 +147,18 @@ getJasmineRequireObj().Suite = function(j$) {
+    return newObj;
+  }
+
+  function clone(obj) {
```
If the `clone` function is no longer used within Suite, we should just remove it.

darthjee replied:
:+1:

slackersoft commented on src/core/Suite.js:
```diff
@@ -88,14 +88,14 @@ getJasmineRequireObj().Suite = function(j$) {
 
   Suite.prototype.clonedSharedUserContext = function() {
-    return clone(this.sharedUserContext());
+    return extend(new j$.UserContext(), this.sharedUserContext());
```
Now that there is a place to put it, could this be a method on the `UserContext` constructor, so Suite would call something like:

```javascript
j$.UserContext.fromExisting(this.sharedUserContext())
```

And remove all of the `extend` stuff from Suite.

darthjee replied:
makes sense ... I didnt want to put in the prototype, but like this would work, will do that later

:+1:

darthjee commented:
@slackersoft after your review, I added more commits for estra specs and changed QueueRunner to never initialize userContext as an empty object.

I am not being able to run the tests on windows, but will have access to a real windows machine on the weekend so I can try running it there
